test/test_xhistogram.cpp

```diff
@@ -29,6 +31,16 @@ namespace xt
             EXPECT_EQ(count(1), 2.);
         }
 
+        {
+            xt::xtensor<double, 1> edges {0, 1.5, 3};
+            xt::xtensor<double, 1> weights {1, 1, 1, 1};
+            auto count = xt::histogram<int>(data, edges, weights, true);
+            // This is indeed a bug. If the count is the density, an integral return
+            // type should not be allowed.
+            EXPECT_EQ(count(0), 0.);
+            EXPECT_EQ(count(1), 0.);
```

Comment:
I would propose to remove the possibility of providing a return type. But this will be a breaking change.

Actually, the behavior of `numpy.histogram` is that:
1. The dtype of count is always `np.int64` if `weights` is not specified and `density` is False;
2. The dtype of count is always `np.float64` if `weights` is given (no matter what dtype weights has) or `density` is True. 

In my opinion, this is also easy to implement.

Here, we could make a similar implementation by making the dtype of the first case to `size_t`, but this will be another breaking change. Or, we could just fix the return type of all cases to `double`.

Reply:
In an ideal world one wants to keep the type of `weights` when it is specified *but* `density = false`. But then is simply no possible I think. To approximate this I think we introduced `R`. Do you remember more @JohanMabille ?

Reply:
Not really, I need to fetch the history to retrieve some details. I'm a bit swamped with other urgent tasks, I will get back to this once it is done (probably next week). Sorry for this delay @zhujun98 and @tdegeus

Reply:
Thanks @JohanMabille! Take your time.

Reply:
It could be good wrap up this PR

Reply:
@JohanMabille Would you have some time for this PR in the coming days? Thank you!

Reply:
Yes I will take the time to do so. BTW, I retrieved the [post](https://github.com/xtensor-stack/xtensor/pull/1108#discussion_r213374492) where we discussed the return type `R`, that was indeed to approximate the behavior that @tdegeus described above.

Reply:
I see your discussion. Anyway, this is something came to my mind which could be improved in the future. For this PR, we should focus on the main topic.